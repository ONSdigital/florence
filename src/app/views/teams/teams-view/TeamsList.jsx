import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router";
import url from "../../../utilities/url";
import Input from "../../../components/Input";
import SimpleSelectableList from "../../../components/simple-selectable-list/SimpleSelectableList";

const propTypes = {
    dispatch: PropTypes.func,
    teams: PropTypes.object.isRequired,
    loadTeams: PropTypes.func.isRequired,
};

const TeamsList = props => {
    const { teams, loadTeams } = props;

    const [isFetchingTeams, setIsFetchingTeams] = useState(true);
    const [filterTerm, setFilterTerm] = useState("");
    const [previewTeams, setPreviewTeams] = useState([]);

    useEffect(() => {
        loadTeams();
    }, []);

    useEffect(() => {
        const listOfPreviewTeams = [];
        teams?.groups?.forEach(team => {
            if (team.group_name.startsWith("role-")){
                return; // role-admin, role-publisher are not preview teams but are stored in groups
            }
            const previewTeamData = {
                row: {
                    title: team.description != null ? team.description : team.group_name,
                    id: team.group_name,
                    url: `./groups/${team.group_name}`,
                    extraDetails: [[{ content: formatDateString(team.creation_date) }]],
                },
                date: team.creation_date, //used for sorting
            };
            listOfPreviewTeams.push(previewTeamData);
            listOfPreviewTeams.sort((a, b) => {
                return new Date(b.date) - new Date(a.date);
            });
        });
        setPreviewTeams(listOfPreviewTeams);
        setIsFetchingTeams(false);
    }, [teams]);

    const formatDateString = date => {
        const utcDate = new Date(date);
        const datePart = utcDate.toLocaleDateString("en-GB", { day: "2-digit", month: "long", year: "numeric" });
        const timePart = utcDate.toLocaleTimeString("en-GB", { hour: "2-digit", minute: "2-digit" });
        const completeFormattedDate = `${datePart} at ${timePart}`;
        return (
            <>
                Created <span className="font-weight--600">{completeFormattedDate}</span>
            </>
        );
    };

    let allRowData = [];
    let rowsToDisplay = [];
    if (previewTeams.length > 0) {
        previewTeams.map(pTeam => {
            allRowData.push(pTeam.row);
        });
        rowsToDisplay =
            filterTerm === ""
                ? allRowData
                : allRowData.filter(
                      team =>
                          team.id.toLowerCase().search(filterTerm.toLowerCase()) !== -1 ||
                          team.title.toLowerCase().search(filterTerm.toLowerCase()) !== -1
                  );
    }

    return (
        <div className="grid grid--justify-space-around">
            <div className="grid__col-9">
                <div className="margin-top--2">
                    &#9664; <Link to={url.resolve("../")}>Back</Link>
                </div>
                <span className="margin-top--1">
                    <h1 className="inline-block margin-top--0 margin-bottom--0 padding-right--1">Preview teams</h1> <Link to={url.resolve("./groups/create")}> Create a new team </Link>
                </span>
                <div className="grid__col-6">
                    <Input
                        id="search-content-types"
                        placeholder="Search teams by name or ID"
                        onChange={e => {
                            e.preventDefault();
                            setFilterTerm(e.target.value);
                        }}
                    />
                </div>
                <SimpleSelectableList rows={rowsToDisplay} showLoadingState={isFetchingTeams} />
            </div>
        </div>
    );
};

TeamsList.propTypes = propTypes;
export default TeamsList;
