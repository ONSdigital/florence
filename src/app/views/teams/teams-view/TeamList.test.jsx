import React from "react";
import renderer from "react-test-renderer";
import {render, screen} from "../../../utilities/tests/test-utils";
import TeamsList from "./TeamsList";
import "@testing-library/jest-dom/extend-expect";

const teams = {
    count: 2,
    groups: [
        {
            creation_date: "2021-10-12T14:32:50.913Z",
            description: "foo",
            group_name: "104ea95f-dcf2-4e36-81cc-bbf840b6ee10",
            last_modified_date: "2021-10-12T14:32:50.913Z",
            precedence: 19,
            role_arn: null,
            user_pool_id: "bar"
        },
        {
            creation_date: "2021-09-14T12:57:50.913Z",
            description: "baz",
            group_name: "45a6c8ed-2494-4e68-89ce-0ca504dc020a",
            last_modified_date: "2021-10-12T12:57:50.238Z",
            precedence: 19,
            role_arn: null,
            user_pool_id: "bar"
        }]
}

const defaultProps = {
    teams: [],
    loadTeams: jest.fn()
};

describe("TeamsList", () => {
    it("matches the snapshot", () => {
        const wrapper = renderer.create(<TeamsList {...defaultProps} />);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it("requests all teams on load", () => {
        render(<TeamsList {...defaultProps} />);
        expect(defaultProps.loadTeams).toHaveBeenCalled();
    });

    it("shows Back Button", () => {
        render(<TeamsList {...defaultProps} />);
        expect(screen.getByText(/Back/i)).toBeInTheDocument();
    });

    it("shows, message if no teams found", () => {
        render(<TeamsList {...defaultProps} />);
        expect(screen.getByText(/Nothing to show/i)).toBeInTheDocument();
    });
    it("shows, creation option and search option", () => {
        render(<TeamsList {...defaultProps} />);
        expect(screen.getByText("Create a new team")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Search teams by name or ID")).toHaveValue("");
    });
    it("shows list of teams details", () => {
        const props = {...defaultProps, teams: teams};
        render(<TeamsList {...props} />);

        expect(screen.getByText(/Back/i)).toBeInTheDocument();
        expect(screen.getByText("Preview teams")).toBeInTheDocument();
        expect(screen.getByText("foo")).toBeInTheDocument();
        expect(screen.getByText("baz")).toBeInTheDocument();
        expect(screen.getByText("12 October 2021 at 15:32")).toBeInTheDocument();
        expect(screen.getByText("14 September 2021 at 13:57")).toBeInTheDocument();
    });
});