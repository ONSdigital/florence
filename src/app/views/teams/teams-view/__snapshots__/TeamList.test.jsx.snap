// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamsList matches the snapshot 1`] = `
<div
  className="grid grid--justify-space-around"
>
  <div
    className="grid__col-9"
  >
    <div
      className="margin-top--2"
    >
      â—€ 
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Back
      </a>
    </div>
    <span
      className="margin-top--1"
    >
      <h1
        className="inline-block margin-top--0 margin-bottom--0 padding-right--1"
      >
        Preview teams
      </h1>
       
      <a
        onClick={[Function]}
        style={Object {}}
      >
         Create a new team 
      </a>
    </span>
    <div
      className="grid__col-6"
    >
      <div
        className="form__input"
      >
        <label
          className="form__label"
          htmlFor="search-content-types"
        />
        <input
          autoComplete="new-password"
          autoFocus={false}
          className="input"
          disabled={false}
          id="search-content-types"
          name="search-content-types"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Search teams by name or ID"
          type="text"
        />
      </div>
    </div>
    <ul
      className="list list--neutral simple-select-list"
    >
      <span
        className="margin-top--1 loader loader--dark"
        data-testid="loader"
      />
      
    </ul>
  </div>
</div>
`;
